autotest/ogr/ogr_sqlite.py

```diff
@@ -3894,6 +3894,181 @@ def test_ogr_sqlite_relationships():
+    assert retrieved_rel.GetType() == gdal.GRT_ASSOCIATION
+    assert retrieved_rel.GetLeftTableName() == "origin_table"
+    assert retrieved_rel.GetRightTableName() == "dest_table"
+    assert retrieved_rel.GetLeftTableFields() == ["o_pkey"]
+    assert retrieved_rel.GetRightTableFields() == ["dest_pkey"]
+    assert retrieved_rel.GetRelatedTableType() == "features"
+    assert retrieved_rel.GetMappingTableName() == ""
+    assert retrieved_rel.GetLeftMappingTableFields() is None
+    assert retrieved_rel.GetRightMappingTableFields() is None
+
+    # try again, should fail because relationship already exists
+    assert not ds.AddRelationship(relationship)
+
+    lyr = ds.CreateLayer("origin_table2", geom_type=ogr.wkbNone)
+    fld_defn = ogr.FieldDefn("o_pkey", ogr.OFTInteger)
+    assert lyr.CreateField(fld_defn) == ogr.OGRERR_NONE
+
+    lyr = ds.CreateLayer("dest_table2", geom_type=ogr.wkbNone)
+    fld_defn = ogr.FieldDefn("dest_pkey", ogr.OFTInteger)
+    assert lyr.CreateField(fld_defn) == ogr.OGRERR_NONE
+
+    # only one-to-many relationships are supported
+    relationship = gdal.Relationship(
+        "my_relationship", "origin_table2", "dest_table2", gdal.GRC_ONE_TO_ONE
+    )
+    relationship.SetLeftTableFields(["o_pkey"])
+    relationship.SetRightTableFields(["dest_pkey"])
+    relationship.SetRelatedTableType("features")
+    assert not ds.AddRelationship(relationship)
+
+    relationship = gdal.Relationship(
+        "my_relationship", "origin_table2", "dest_table2", gdal.GRC_MANY_TO_MANY
+    )
+    relationship.SetLeftTableFields(["o_pkey"])
+    relationship.SetRightTableFields(["dest_pkey"])
+    relationship.SetRelatedTableType("features")
+    assert not ds.AddRelationship(relationship)
+
+    relationship = gdal.Relationship(
+        "my_relationship", "origin_table2", "dest_table2", gdal.GRC_MANY_TO_ONE
+    )
+    relationship.SetLeftTableFields(["o_pkey"])
+    relationship.SetRightTableFields(["dest_pkey"])
+    relationship.SetRelatedTableType("features")
+    assert not ds.AddRelationship(relationship)
+
+    # add second relationship of composition type
+    lyr = ds.CreateLayer("origin_table3", geom_type=ogr.wkbNone)
+    fld_defn = ogr.FieldDefn("o_pkey", ogr.OFTInteger)
+    assert lyr.CreateField(fld_defn) == ogr.OGRERR_NONE
+
+    lyr = ds.CreateLayer("dest_table3", geom_type=ogr.wkbNone)
+    fld_defn = ogr.FieldDefn("dest_pkey", ogr.OFTInteger)
+    assert lyr.CreateField(fld_defn) == ogr.OGRERR_NONE
+
+    relationship = gdal.Relationship(
+        "my_relationship", "origin_table3", "dest_table3", gdal.GRC_ONE_TO_MANY
+    )
+    relationship.SetType(gdal.GRT_COMPOSITE)
+
+    relationship.SetLeftTableFields(["o_pkey"])
+    relationship.SetRightTableFields(["dest_pkey"])
+    ds.ExecuteSQL(
+        "CREATE UNIQUE INDEX origin_table3_o_pkey_idx ON origin_table3(o_pkey)"
+    )
+
+    assert ds.AddRelationship(relationship)
+
+    assert set(ds.GetRelationshipNames()) == {
```

Comment:
before adding the relationship, it could be interesting to add a feature in the table that gets foreign key constraints added and read it back after AddRelationship() to check the table was recreated OK. And perhaps do that after dataset closing and re-opening

Reply:
Do you mean a feature which would violate the foreign key constraint?

Reply:
no, just a regular feature, to test that field recreation logic of AddForeignKeysToTable() does its job right. Currently I guess that if for example it was bogus and would just wipe out completely the table, that would be unnoticed by the current test.

Reply:
Ah, gotcha :+1: